src/predictor/cpu_predictor.cc

```diff
@@ -99,52 +140,32 @@ class AdapterView {
   bst_row_t const static base_rowid = 0;  // NOLINT
 };
 
-template <typename DataView>
-void PredictBatchKernel(DataView batch, std::vector<bst_float> *out_preds,
-                        gbm::GBTreeModel const &model, int32_t tree_begin,
-                        int32_t tree_end,
-                        std::vector<RegTree::FVec> *p_thread_temp) {
+template <typename DataView, size_t block_of_rows_size>
+void PredictBatchByBlockOfRowsKernel(DataView batch, std::vector<bst_float> *out_preds,
+                                     gbm::GBTreeModel const &model, int32_t tree_begin,
+                                     int32_t tree_end,
+                                     std::vector<RegTree::FVec> *p_thread_temp) {
   auto& thread_temp = *p_thread_temp;
+  const std::vector<std::unique_ptr<RegTree>> &trees = model.trees;
   int32_t const num_group = model.learner_model_param->num_output_group;
 
-  std::vector<bst_float> &preds = *out_preds;
   CHECK_EQ(model.param.size_leaf_vector, 0)
       << "size_leaf_vector is enforced to 0 so far";
   // parallel over local batch
   const auto nsize = static_cast<bst_omp_uint>(batch.Size());
-  auto constexpr kUnroll = DataView::kUnroll;
-  const bst_omp_uint rest = nsize % kUnroll;
-  if (nsize >= kUnroll) {
-#pragma omp parallel for schedule(static)
-    for (bst_omp_uint i = 0; i < nsize - rest; i += kUnroll) {
-      const int tid = omp_get_thread_num();
-      RegTree::FVec &feats = thread_temp[tid];
-      int64_t ridx[kUnroll];
-      SparsePage::Inst inst[kUnroll];
-      for (size_t k = 0; k < kUnroll; ++k) {
-        ridx[k] = static_cast<int64_t>(batch.base_rowid + i + k);
-      }
-      for (size_t k = 0; k < kUnroll; ++k) {
-        inst[k] = batch[i + k];
-      }
-      for (size_t k = 0; k < kUnroll; ++k) {
-        for (int gid = 0; gid < num_group; ++gid) {
-          const size_t offset = ridx[k] * num_group + gid;
-          preds[offset] += PredValue(inst[k], model.trees, model.tree_info, gid,
-                                     &feats, tree_begin, tree_end);
-        }
-      }
-    }
-  }
-  for (bst_omp_uint i = nsize - rest; i < nsize; ++i) {
-    RegTree::FVec &feats = thread_temp[0];
-    const auto ridx = static_cast<int64_t>(batch.base_rowid + i);
-    auto inst = batch[i];
-    for (int gid = 0; gid < num_group; ++gid) {
-      const size_t offset = ridx * num_group + gid;
-      preds[offset] += PredValue(inst, model.trees, model.tree_info, gid,
-                                 &feats, tree_begin, tree_end);
-    }
+
+  const bst_omp_uint n_row_blocks = (nsize) / block_of_rows_size + !!((nsize) % block_of_rows_size);
+#pragma omp parallel for schedule(guided)
+  for (bst_omp_uint block_id = 0; block_id < n_row_blocks; ++block_id) {
+    const size_t batch_offset = block_id*block_of_rows_size;
+    const size_t block_size = std::min(nsize - batch_offset, block_of_rows_size);
+    const size_t fvec_offset = omp_get_thread_num()*block_of_rows_size;
```

Comment:
```suggestion
    const size_t fvec_offset = omp_get_thread_num() * block_of_rows_size;
```

Use spaces around the operator.